pex/layout.py

```diff
@@ -126,18 +149,243 @@ def record(self, dest_dir):
+                        ) as spread_chroot:
+                            if not spread_chroot.is_finalized():
+                                with TRACER.timed({extracting_msg!r}):
+                                    layout.extract_dist(
+                                        dest_dir=spread_chroot.work_dir,
+                                        dist_relpath={dist_relpath!r},
+                                        is_wheel_file={is_wheel_file!r}
+                                    )
+
+                        symlink_dest = {symlink_dest!r}
+                        safe_mkdir(os.path.dirname(symlink_dest))
+                        os.symlink({symlink_src!r}, symlink_dest)
+                        """
+                    ).format(
+                        pex=layout.path,
+                        spread_dest=spread_dest,
+                        source=source,
+                        extracting_msg=extracting_message,
+                        dist_relpath=dist_relpath,
+                        is_wheel_file=deps_are_wheel_files,
+                        symlink_src=symlink_src,
+                        symlink_dest=symlink_dest,
+                    ),
+                ],
+                expose=["pex"],
+            ),
+            result=None,
+        )
+
+    # Assuming that extract / install time scales with distribution size, we ensure no job slot is
+    # so unlucky as to get all the biggest jobs and thus become an un-necessarily long pole by
+    # sorting based on distribution size. Some examples to illustrate the effect using 6 input
+    # distributions and 2 job slots:
+    #
+    # 1.) Random worst case ordering:
+    #         [9, 1, 1, 1, 1, 10] -> slot1[9] slot2[1, 1, 1, 1, 10]: 14 long pole
+    #     Sorted becomes:
+    #         [10, 9, 1, 1, 1, 1] -> slot1[10, 1, 1] slot2[9, 1, 1]: 12 long pole
+    # 2.) Random worst case ordering:
+    #         [6, 4, 3, 10, 1, 1] -> slot1[6, 10] slot2[4, 3, 1, 1]: 16 long pole
+    #     Sorted becomes:
+    #         [10, 6, 4, 3, 1, 1] -> slot1[10, 3] slot2[6, 4, 1, 1]: 13 long pole
+    #
+    # TODO(John Sirois): Consider having execute_parallel take an optional costing function and
+    #  move this sorting logic and explanation centrally there.
+    inputs = sorted(
+        (item for item in pex_info.distributions.items()),
+        key=lambda item: layout.dist_size(
+            os.path.join(DEPS_DIR, item[0]), is_wheel_file=deps_are_wheel_files
+        ),
+        reverse=True,
+    )
+    with TRACER.timed(
+        "Using a maximum of {max_jobs} parallel jobs to install {count} distributions".format(
+            max_jobs="<auto>" if max_jobs == 0 else max_jobs, count=len(pex_info.distributions)
+        )
+    ):
+        for _ in execute_parallel(
+            inputs=inputs, spawn_func=install_distribution, max_jobs=max_jobs
+        ):
+            pass
+
+
+AVERAGE_DISTRIBUTION_SIZE_PARALLEL_JOB_THRESHOLD = 1 * 1024 * 1024  # ~1MB
```

Comment:
Does this number come from experiments?

Reply:
Yes. As noted elsewhere, on 1 machine and with limited cases. I'm very much not confident in the heuristic.

Reply:
Commented.